<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../js/jquery.ba-hashchange.min.js"></script>
		<script type="text/javascript" src="../js/local.js"></script>

		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>

		<script type="text/javascript">
			$(document).ready(
				function() {
					setFirstPage( "overview" ) ;
				}
			) ;
		</script>
		<link href="../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
			<h2>Obtaining QCTOOL</h2>
				<p>
					QCTOOL is available either as binaries or as source code.  The recommended approach is to download and compile
					the QCTOOL source code on the machine(s) you will be working on.  However, for convenience a small number of
					precompiled binary packages are provided below.   These provided as-is and will not work on all platforms; if a suitable version is not available please
					download the QCTOOL source code and compile instead.
				</p>
				<table class="file_formats">
					<tr><th>Version</th><th>Platform</th><th>file</th>
					<tr><td>v2.0.6</th><th>Mac OS X</th><th><a href="https://www.well.ox.ac.uk/~gav/resources/qctool_v2.0.6-osx.tgz">qctool_v2.0.6-osx.tgz</a></th>
					<tr><td>v2.0.6</th><th>Ubuntu Linux 16.04</th><th><a href="https://www.well.ox.ac.uk/~gav/resources/qctool_v2.0.6-Ubuntu16.04-x86_64.tgz">qctool_v2.0.6-Ubuntu16.04-x86_64.tgz</a></th>
					<tr><td>v2.0.6</th><th>CentOS Linux 7.3</th><th><a href="https://www.well.ox.ac.uk/~gav/resources/qctool_v2.0.6-CentOS_Linux7.3.1611-x86_64.tgz">qctool_v2.0.6-CentOS_Linux7.3.1611-x86_64.tgz</a></th>
				</table>
				</table>
				<h3>
					Downloading source
				</h3>
				<p>
					The source code to qctool is available as a <a href="https://fossil-scm.org">fossil</a> repository
					at <a href="https://code.enkre.net/qctool">https://code.enkre.net/qctool</a>.
					To obtain source code, use the fossil command-line client, or alternatively download a
					tarball from this table.  (Note that this currently generates a tarball on the fly, which may take a while).
				</p>
				<table class="file_formats">
					<tr><th>Version</th><th>file</th>
					<tr><td>release</th><th><a href="https://code.enkre.net/qctool/zip/v2.0.7/qctool.tgz">https://code.enkre.net/qctool/zip/release/qctool.tgz</a></th>
					<tr><td>trunk</th><th><a href="https://code.enkre.net/qctool/zip/trunk/qctool.tgz">https://code.enkre.net/qctool/zip/trunk/qctool.tgz</a></th>
				</table>
				<h3>
					Versioning scheme
				</h3>
				<p>
					As of version 2.1, QCTOOL will follow a versioning scheme whereby odd minor numbers (2.1.x etc.) represent
					development work while even numbers (2.0.x) represent release versions.
					The current latest release versions are 2.0 version.
				</p>
				<h3>
					Compiling
				</h3>
				<p>
					Having downloaded the source code (and extracted it if necessary), a basic compilation sequence would then be:
				</p>
<pre>
$ cd qctool
$ ./waf-1.5.18 configure
$ ./waf-1.5.18
</pre>
				<p>
					If all is successful, this produces an executable
					<code>./build/release/qctool_v2.0-&lt;version&gt;</code>
				</p>
				</div>
			<p>
				You can check qctool is working by running it:
			</p>
			<pre>
$ ./build/release/qctool_v2.0-&lt;version&gt; -help
</pre>
			<p>which should print out a list of options.</p>
			<h3>Further information</h3>
			<p>
				The <a href="https://code.enkre.net/qctool/wiki?name=Home">QCTOOL wiki</a> contains further information
				which may be useful if you are unable to compile QCTOOL or need to tweak compilation for specific platforms.
			</p>
		</section>

		<nav class="button_bar"><div>
			<div class="nav_button" name="overview">
				<a href="../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../documentation/alphabetical_options.html">List of options</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../documentation/examples/basic.html">Basic usage</a></li>
					<li><a href="../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../documentation/examples/converting_genotypes.html">Thresholding genotypes</a></li>
					<li><a href="../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../documentation/examples/combining.html">Combining datasets</a></li>
					<li><a href="../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../documentation/examples/filtering_samples.html">Filtering samples</a></li>
					<li><a href="../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../documentation/examples/annotating.html">Computing annotations</a></li>
					<li><a href="../documentation/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
					<li><a href="../documentation/examples/computing_ld.html">Computing LD</a></li>
					<li><a href="../documentation/examples/computing_pcs.html">Computing principal components</a></li>
					<li><a href="../documentation/examples/inferring_ploidy.html">Working with sex chromosomes</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../documentation/download.html">download</a>
			</div>
		</div>
</nav>
		
	</body>
</html>
