<section class="main_text">
			<h2>Converting between file formats</h2>
		<div class="task">
			<div class="task_name">
				Basic conversions
			</div>
			<div class="task_notes">
				The basic format of a conversion command is:
			</div>
			<div class="task_command_line">
				$ qctool -g &lt;input file&gt; [-filetype &lt;input type&gt;] -og &lt;output file&gt; [-ofiletype &lt;output type&gt;] [+modifier options]
			</div>
			<div class="task_notes">
				QCTOOL normally deduces file types from the file extension.
				However, for filetypes that are not automatically recognised, or if you want to specify,
				the <code>-filetype</code> and <code>-ofiletype</code> options can be used to specify the types.
				The <a href="../../documentation/genotype_file_formats.html">genotype file formats page</a> lists file type specifiers and
				any applicable modifier options.
			</div>
			<div class="task_notes">
				E.g. to convert between between bgen and vcf formats:
			</div>
			<div class="task_command_line">
				$ qctool -g example.bgen -og example.vcf
			</div>
		</div>
		<div class="task">
			<div class="task_name">
				Handling files split by chromosome
			</div>
			<div class="task_notes">
				If the input filename contains a # character, e.g. <code>example_#.gen</code> this is treated as a
				chromosomal wildcard and will match all (human) chromosomes. This will also be used to
				infer the chromosome for each variant if chromosome information is not present in the files themselves.
				For example, the command:
			</div>
			<div class="task_command_line">
				$ qctool -g example_#.gen -og example.bgen
			</div>
			<div class="task_notes">
				will process all 22 example files, and if the input GEN files have no chromosome information included,
				the output data will have chromosome identifiers taken from the filenames.
			</div>
		</div>
		<div class="task">
			<div class="task_name">
				Including sample information in conversions
			</div>
			<div class="task_notes">
				 Although specifying a sample file in conversions is optional, it's sometimes preferable to include one so that output files have
				the correct sample identifiers (e.g. when converting GEN format to BGEN or VCF), e.g.:
			</div>
			<div class="task_command_line">
				$ qctool -g example_#.gen -og example.bgen -s example.sample
			</div>
			<div class="task_notes">
				which will produce a BGEN file with a sample identifier block.
			</div>
		</div>
		<div class="task">
			<div class="task_name">
				Converting from VCF format
			</div>
			<div class="task_notes">
				By default QCTOOL reads genotype calls from the <span class="monospace">GT</span> field in the VCF file.
				The <code>-vcf-genotype-field</code> can be used to alter this, e.g.:
			</div>
			<div class="task_command_line">
				$ qctool -g example.vcf -vcf-genotype-field GP -og converted.bgen
			</div>
			<div class="task_notes">
				QCTOOL also assumes that VCF metadata is correct (according to the VCF v4.2 spec)- and it will
				fail with an error if that's not the case.
				To work around this, the <code>-metadata</code> option can be used to tell QCTOOL to load metadata from an
				external file:
			</div>
			<div class="task_command_line">
				$ qctool -g example.vcf -og converted.bgen -metadata metadata.txt
			</div>
			<div class="task_notes">
				The specified metadata file should contain complete VCF metadata, starting
				with the <code>##fileformat=VCFv4.2</code> line, up to but not
				including the <code>#CHROM...</code> line.
			</div>
		</div>
	</div>
</section>
