<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="js/jquery.ba-hashchange.min.js"></script>
		<script type="text/javascript" src="js/local.js"></script>

		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>

		<script type="text/javascript">
			$(document).ready(
				function() {
					setFirstPage( "overview" ) ;
				}
			) ;
		</script>
		<link href="style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>
		<div class="button_bar">
			<div class="nav_button" name="overview">
				<a href="index.html" class="nav_button_content">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				<a href="reference.html" class="nav_button_content">documentation</a>
			</div>
			<div class="nav_button" name="tutorial">
				<a href="examples.html" class="nav_button_content">examples</a>
			</div>
			<div class="nav_button" name="file_formats">
				<a href="file_formats.html" class="nav_button_content">file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="download.html" class="nav_button_content">download</a>
			</div>
		</div>
		<div class="main_text">
			<div id="file_formats">
				<div>
					QCTOOL supports the following file formats for genotype data:
					<table class="file_formats">
						<tr>
							<th>
								format
							</th>
							<th>
								recognised extension(s)
							</th>
							<th>
								notes
							</th>
						</tr>
						<tr>
							<td>
								<a href="http://www.stats.ox.ac.uk/%7Emarchini/software/gwas/file_format.html">GEN format</a>
							</td>
							<td>
								<span class="monospace">.gen[.gz]</span>
							</td>
							<td>
								Optionally, an extra initial column containing chromosomes can be included in the input.  QCTOOL auto-detects this
								in input files by counting the columns in the file.
								To suppress this column in output files, use the <span class="monospace">-omit-chromosome</span> option.
							</td>
						</tr>
						<tr>
							<td>
								<a href="http://www.well.ox.ac.uk/~gav/bgen_format/bgen_format.html">BGEN format</a>
							</td>
							<td>
								<code>.bgen</code>
							</td>
							<td>
								Output files are in BGEN v1.3 with 16 bits per probability and compressed
								using <a href="http://www.zstd.net">zstandard</a> by default. 
								The <code>-bgen-bits</code> option can be used to adjust the number of bits used.
								The <code>-bgen-compression</code> option can be used to choose the method of compression
								(zlib or zstandard).  Use <code>-ofiletype bgen_v1.1</code> to specify
								files compatible with the BGEN v1.1 spec (these always use 16 bits per probability and zlib compression).
							</td>
						</tr>
						<tr>
							<td>
								<a href="http://samtools.github.io/hts-specs/VCFv4.2.pdf">Variant call format</a>
							</td>
							<td>
								<span class="monospace">.vcf[.gz]</span>
							</td>
							<td>
								QCTOOL is strict about metadata in input files for the fields it reads;
								since this is not always correct a <span class="monospace">-metadata</span> option is provided to override
								the input file metadata. Currently, only genotypes are output when outputting VCF files.
								Note that QCTOOL does not apply PHRED scaling to probabilities in the GP field.
							</td>
						</tr>
						<tr>
							<td>
								<a href="">PLINK binary (bed) format</a>
							</td>
							<td>
								<span class="monospace">.bed</span>
							</td>
							<td>
								Use <code>-[o]filetype binary_ped</code> to read/write this format.  Note that QCTOOL currently does only the most basic processing of
								FAM files: when reading, it uses them to count the number of samples in the BED file, when writing it writes a FAM file
								with missing data in all fields except the ID field.  You will therefore need to create fuller FAM files seperately for use with
								other tools.
							</td>
						</tr>
						<tr>
							<td>
								<a href="http://www.shapeit.fr/">SHAPEIT</a> haplotype format
							</td>
							<td>
								
							</td>
							<td>
								Use <span class="monospace">-[o]filetype shapeit_haplotypes</span> to specify reading or writing this file type.
								For genotypic computations, genotypes are formed from pairs of haplotypes; it is assumed that the two haplotypes for each individual
								are consecutive columns in the file. (This format is
								described <a href="http://www.shapeit.fr/pages/hapssample.html">here</a>.)
							</td>
						</tr>
						<tr>
							<td>
								<a href="http://mathgen.stats.ox.ac.uk/impute/impute_v2.html">IMPUTE</a> allele probabilities format
							</td>
							<td>

							</td>
							<td>
								Use <span class="monospace">-[o]filetype impute_allele_probs</span>
								to specify reading or writing this filetype.  This file format is like the
								shapeit haplotype format but contains a probability for
								each haplotype (i.e. two probabilities per individual),
								specifying the probability that the haplotype carries the second allele.
							</td>
						</tr>
						<tr>
							<td>
								<a href="http://mathgen.stats.ox.ac.uk/impute/impute_v2.html">IMPUTE</a> haplotype format
							</td>
							<td>
								
							</td>
							<td>
								<b>Input only</b>. To specify this filetype, use <span class="monospace">-filetype impute_haplotypes</span>.
								It is assumed the legend file name is the same as the haplotypes file name, minus extension,
								with <span class="monospace">.legend</span> appended; QCTOOL will also remove/add the
								<code>.gz</code> extension as appropriate.
								For genotypic computations, genotypes are formed from pairs of haplotypes;
								it is assumed that the two haplotypes for each individual are consecutive columns in the haplotypes file.
							</td>
						</tr>
						<tr>
							<td>
								<a href="https://oxfordhla.well.ox.ac.uk/hla/">HLAIMP</a> probability format
							</td>
							<td>
								
							</td>
							<td>
								<b>Input only</b>.	 Use <span class="monospace">-filetype hlaimp</span> to specify reading this file type.
								Currently, this input format implicitly splits each HLA locus as a series of bi-allelic variants.
							</td>
						</tr>
						<tr>
							<td>
								<a href="http://penncnv.openbioinformatics.org/en/latest/">PennCNV</a>/<a href="https://sites.google.com/site/quantisnp/">QuantiSNP</a> format
							</td>
							<td>
								
							</td>
							<td>
								<b>Output only</b>; use <span class="monospace">-ofiletype penncnv</span> to specify writing this file type.
								PennCNV uses a single sample per input file, this can be acheived using the sample filtering options,
								e.g. <code>-incl-samples-where ID_1=&lt;identifier&gt;</code>
							</td>
						</tr>
						<tr>
							<td>
								QCTOOL 'long' format
							</td>
							<td>
								
							</td>
							<td>
								<b>Input only</b>; use <span class="monospace">-filetype long</span> to specify reading this file type.
								Input must be a file with columns <code>SNPID, rsid, chromosome, position, number_of_alleles,
									allele1, other_alleles, sample_id, ploidy, genotype</code>.
								Further columns may also be included (but QCTOOL ignores these).
								Allelesin the other_alleles column must be comma-separated (as with <a href="http://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF</a> ALT alleles).
								When outputting to vcf format, both genotype (GT) and a field 'typed' indicating whether a row for each
								sample and variant was present will be output.
							</td>
						</tr>
						<tr>
							<td>
								<a href="http://stephenslab.uchicago.edu/software.html#bimbam">BIMBAM</a> dosage format; QCTOOL dosage format
							</td>
							<td>
								<code>.dosage[.gz]</code>,
							</td>
							<td>
								<b>Output only</b>; use <code>-ofiletype bimbam_dosage</code>
								or <code>-ofiletype dosage</code> to specify writing this file type.
								This file outputs a single column per sample (named by the sample identifier) containing the
								expected second allele dosage for the sample at each variant.
								The formats differ in that BIMBAM format has no chromosome/position information.
							</td>
						</tr>
						<tr>
							<td>
								QCTOOL intensity text format
							</td>
							<td>
								<code>.intensity[.gz]</code>
							</td>
							<td>
								<b>Output only</b>; use <code>-ofiletype intensity</code> to specify writing this file type.
								The output file has two columns per sample, representing X and Y channel intensities for the
								sample at each variant.  Currently data must be read from a VCF file; the field is specified
								using the <code>-vcf-intensity-field</code> option.
							</td>
						</tr>
					</table>
				</div>
			</div>
			<div id="download">
				<div>
					<p>
						QCTOOL is available either as binaries or as source code.
					</p>
					
					<h3>
						Binaries
					</h3>
					<p>
						Pre-compiled binaries of the QCTOOL v2 release candidate can be found <a href="http://www.well.ox.ac.uk/~gav/resources/">here</a>.
						These are currently provided as-is.  If a version suitable for your platform is not available please
						download QCTOOL and compile instead.
					</p>
					<h3>
						Installing from source
					</h3>
					</details>
					<p>
						The source code to qctool is available as a mercurial repository hosted <a href="http://www.bitbucket.org/gavinband/qctool">on bitbucket</a>.
						Assuming you have <a href="">mercurial</a> installed, the source for the qctool v2 beta branch
						can be downloaded using the command
					</p>
					<div class="terminal">
						<code>$ hg clone --rev beta https://gavinband@bitbucket.org/gavinband/qctool</code>
					</div>
<p>
	Alternatively, a tarball of the latest qctool v2 beta source code can be downloaded
	from <a href="https://bitbucket.org/gavinband/qctool/downloads?tab=branches">the download page on bitbucket</a>.
	Make sure to select the 'beta' branch for qctool v2.  You will need to extract the tarball using a command like
	<code>tar -xzf &lt;filename&gt;.tgz</code>.
</p>
<p>
	Having downloaded the source code (and extracted it if necessary), a basic compilation sequence would then be:
	<pre>
	$ cd qctool
	$ ./waf-1.5.18 configure
	$ ./waf-1.5.18
</pre>
</p>
					</div>
					<p>
						This produces an executable
					<code>./build/release/qctool-v2.0-&lt;version&rt;</code>
				</p>
				<p>
					(Note that QCTOOL relies on <a href="http://www.zlib.net">zlib</a> which must be installed (with headers) 
					on the system before compilation; the configure step will fail if this is not found.)
				</p>
				<p>
					You can check qctool is working by running it:
				</p>
				<pre>
				$ ./build/release/qctool-v2.0-&lt;version&rt; -help
</pre>
<p>This should print out a list of options.</p>
					<p>
					<b>Note:</b> a debug build is also created in <code>./build/default/qctool_v2.0-&lt;version&rt;</code>, use this if you want
					to run qctool in a debugger.
				</p>
				</div>
			</div>
		</div>
	</body>
</html>
