<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" async
		  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML">
		</script>
		<script type="text/javascript" src="../js/jquery-3.2.1.min.js"></script>
		<script type="text/javascript" src="../js/datatables.min.js"></script>
		<script type="text/javascript" src="../js/local.js"></script>

		<link href="../style/css/datatables.min.css" rel="stylesheet" type="text/css">
		<link href="../style/css/default.css" rel="stylesheet" type="text/css">
		<title>HPTEST</title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				HPTEST
			</div>
			<div class="header_subtext">
				Software to test for genetic association between hosts and pathogens.
			</div>
		</div>

<section class="main_text">
<p>HPTEST fits the following model:</p>
<p>
<script type="math/tex; mode=display">\text{log-odds}( \text{outcome genotype} = 1 ) = X^t \cdot \beta</script>
</p>
<p>where <script type="math/tex">X</script> is a function of the predictor genotype and covariates,</p>
<p>
<script type="math/tex; mode=display">X = \left( \begin{matrix} 1 \\ F(\text{predictor genotype}) \\ Z \end{matrix} \right)</script>
</p>
<p>and <script type="math/tex">\beta</script> is a vector of regression coefficients to be maximised over. The full likelihood is
obtained as the product of the above over all samples in the data.</p>
<p>HPTEST currently assumes that both outcome and predictor variants are biallelic (or at least that
they have been split into biallelic records before running through HPTEST). The outcome and
predictor genotypes are measured with respect to those alleles, i.e. the effect is always
interpreted as effect of predictor allele 2 on outcome allele 2. If the outcome has ploidy larger
than 1, then its alleles are regarded as forming independent draws from the Bernoulli distribution
modelled above.</p>
<p>The predictor genotype is currently assumed to be diploid. To handle different possible modes of
inheritance, HPTEST uses a variety of function <script type="math/tex">F</script> above as follows. If <script type="math/tex">g \in \{0,1,2\}</script> denotes
the predictor genotype then the following functions <script type="math/tex">F</script> are used:</p>
<table>
<thead>
<tr>
<th>Mode of inheritance</th>
<th>F</th>
<th>Specified by</th>
</tr>
</thead>
<tbody>
<tr>
<td>additive</td>
<td>
<script type="math/tex">F(g) = g</script>
</td>
<td><code>-model add</code></td>
</tr>
<tr>
<td>dominant</td>
<td>
<script type="math/tex">F(g) = 1</script> if <script type="math/tex">g > 0</script>, or <script type="math/tex">0</script> otherwise</td>
<td><code>-model dom</code></td>
</tr>
<tr>
<td>recessive</td>
<td>
<script type="math/tex">F(g) = 1</script> if <script type="math/tex">g = 2</script>, or <script type="math/tex">0</script> otherwise</td>
<td><code>-model rec</code></td>
</tr>
<tr>
<td>heterozygote</td>
<td>
<script type="math/tex">F(g) = 1</script> if <script type="math/tex">g = 1</script>, or <script type="math/tex">0</script> otherwise</td>
<td><code>-model het</code></td>
</tr>
<tr>
<td>general</td>
<td>
<script type="math/tex">F(g) = (g,1)</script> if <script type="math/tex">g = 1</script>, or <script type="math/tex">(g,0)</script> otherwise</td>
<td><code>-model add+het</code></td>
</tr>
</tbody>
</table>
</section>
		<nav class="button_bar">
			<div>
				<div class="nav_button" name="overview">
					<a href="../index.html">overview</a>
				</div>
				<div class="nav_button" name="documentation">
					documentation
					<ul class="menu">
						<li><a href="../documentation/getting_started.html">Getting started</a></li>
						<li><a href="../documentation/alphabetical_options.html">List of options</a></li>
						<li><a href="../documentation/specifying_priors.html">Specifying priors</a></li>
						<li><a href="../documentation/interpreting_hptest_output.html">Interpreting HPTEST output</a></li>
						<li><a href="../documentation/filtering.html">Filtering variants and samples</a></li>
						<li><a href="../documentation/technical_details.html">Technical details</a></li>
					</ul>
				</div>
				<div class="nav_button" name="download">
					<a href="../documentation/download.html">download</a>
				</div>
			</div>
		</nav>

	</body>
</html>
