<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../../js/local.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="../../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
			<h2>Altering variant identifying data</h2>
					<div class="task_command_line">
						$ qctool -g &lt;input file(s)&gt; -strand &lt;strand file&gt; [-flip-to-match-allele &lt;column name&gt;] ...
					</div>
					<div class="task_notes">
						The <code>-strand</code> option tells QCTOOL to read strand information from the specified file and apply strand alignment
						to the data.  Alleles at variants where the strand is '+' will be processed unchanged; alleles
						at variants where the strand is '-' will be complemented (i.e. A&lt;-&gt;T, G&lt;-&gt;C);
						alleles at variants which have missing strand information will be output in a way which records the ambiguous strand
						(and consequently are probably not useful for downstream analyses).
						Strand files should have six named columns,
						<code>SNPID</code>,
						<code>rsid</code>,
						<code>chromosome</code>,
						<code>position</code>,
						<code>alleleA</code>,
						<code>alleleB</code>,
						<code>strand</code>; plus any additional columns.
					</div>
					<div class="task_notes">
						If the <code>-flip-to-match-allele</code> option is given, the strand file must contain a column with the given name.
						Alleles and genotypes are then also recoded so that the allele in the specified
						column is the first allele and the other allele the second allele.
					</div>
				</div>
		
				<div class="task">
					<div class="task_name">
						Threshhold genotype probabilities (e.g. GP to GT)
					</div>
					<div class="task_command_line">
						$ qctool -g input.gen -og example.vcf -threshhold 0.9
					</div>
					<div class="task_notes">
						Threshholding takes genotype probabilities (specified in a GEN or BGEN file, or GP field of a VCF file)
						and produces hard calls (e.g. GT field in a VCF file).  Thus, the output file from the above command will have a <code>FORMAT</code>
						of <code>GT:GP</code> for each SNP.
					</div>
					<div class="task_notes">
						In general <code>-threshhold</code> makes the threshholded call the primary genotype field for analysis; thus it
						changes the output of <code>-snp-stats</code> and other computations below.
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Manipulate chromosome information
					</div>
					<div class="task_command_line">
						$ qctool -g example_01.gen -og example_01.bgen -assume-chromosome 01
					</div>
					<div class="task_notes">
						This converts a genotypes file, filling in the chromosome information if it is not present in the input data.
					</div>
					<div class="task_command_line">
						$ qctool -g example_01.bgen -og example_01.gen -omit-chromosome
					</div>
					<div class="task_notes">
						This writes GEN-format files without a chromosome column.	These are suitable for use with other programs such as GTOOL and IMPUTE.
						(See the <a href="../../documentation/file_formats.html">file formats page</a> for a description of file formats understood by QCTOOL.)
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Use qctool in a pipeline
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -og - | less -S
					</div>
					<div class="task_notes">
						Note: by default this outputs genotypes in GEN format, use <code>-ofiletype</code> to alter this.
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -snp-stats -osnp - | less -S
					</div>
					<div class="task_notes">
						Note: currently this outputs genotypes in GEN format.
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Filter samples
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen 
						-s example.sample 
						-og filtered.bgen
						-excl-samples samples.txt
					</div>
					<div class="task_notes">
						These commands filters out / filter in all samples with identifier in the file <code>samples.txt</code>, which should contain a whitespace-separated list of
						sample identifiers to exclude.  Samples are identified by the first identifier field (often <code>ID_1</code>) in the sample file, or if not
						sample file is specified, by sample identifiers specified in genotype data source (e.g. by the header in vcf or bgen formats).
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen 
						-s example.sample 
						-og filtered.bgen
						-incl-samples samples.txt
					</div>
					<div class="task_notes">
						Ditto, but this filters out all samples <em>not</em> in the file <code>samples.txt</code>.
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen 
						-s example.sample 
						-og filtered.bgen
						-excl-samples-where &lt;column&gt;=&lt;value&gt;
					</div>
					<div class="task_notes">
						This command excludes all samples which have the specified value in the named column
						of the sample file.  For example, to exclude males, <code>-incl-samples-where &lt;sex=M&gt;</code>
						might be used.  To instead exclude samples with all but a specified value use:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen 
						-s example.sample 
						-og filtered.bgen
						-excl-samples-where &lt;column&gt;!=&lt;value&gt;
					</div>
					<div class="task_notes">
						Examples of the use of this might be: excluding males from a sex chromosome analysis (e.g. <code>-excl-samples-where sex=M</code>);
						excluding specific individuals (e.g. <code>-excl-samples-where ID_1=&lt;something&gt;</code>); excluding
						individuals for which principal components are not computed (e.g. <code>-excl-samples-where PC1=NA</code>.)
					</div>
					<div class="task_notes">
						It's also possible to specify an inclusion criterion:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen 
						-s example.sample 
						-og filtered.bgen
						-incl-samples-where &lt;column&gt;=&lt;value&gt;
					</div>
					<div class="task_notes">
						In general inclusion and exclusion options come in pairs of the form <code>-incl-&lt;option&gt;</code>, <code>-excl-&lt;option&gt;</code>.
						If multiple overlapping conditions are specified, the logic is that samples that satisfy any of the specified exclusion criteria will
						be excluded, and likewise samples that do not satisfy any of the specified inclusion criteria will be excluded.
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Filter variants
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -og subsetted.bgen -excl-rsids rsids_to_exclude.txt
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -og subsetted.bgen -excl-snpids snpids_to_exclude.txt
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -og subsetted.bgen -excl-positions positions_to_exclude.txt
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -og subsetted.bgen -excl-range 06:25000000-40000000
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -og subsetted.bgen -excl-variants snps_to_exclude.txt
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -og subsetted.bgen -excl-variants-matching rsid~rs1%
					</div>
					<div class="task_notes">
						Note: inclusion/exclusion options come in pairs named <code>-excl-&lt;option&gt;</code> / <code>-incl-&lt;option&gt;</code>.
						A variant will be excluded if it matches any exclusion criterion.  If inclusion criteria are specified, variants are also excluded
						if they do not match any inclusion criterion.
					</div>
					<div class="task_notes">
						The file passed to <span class="monospace">-[in|ex]cl-positions</span> should contain a whitespace-separated list of positions in the form <span class="monospace">&lt;chromosome&gt;:&lt;position&gt;</span>.
					</div>
					<div class="task_notes">
						The file passed to <span class="monospace">-[in|ex]cl-snps</span> should have
						six (named) columns: SNPID, rsid, chromosome, position, first and second alleles.
					</div>
					<div class="task_notes">
						The argument to <span class="monospace">-[in|ex]cl-snps-matching</span> specifies a
						pattern to use to match SNPs.  A limited wildcard functionality is implemented:
						if the last character of the argument is %, this stands for a wildcard and will match
						any string. Thus e.g. <code>-excl-variants-matching rsid~rs1%</code>
						will exclude any SNP whose primary identifier begins with <code>rs1</code>.  It's also
						possible to match by alternate identifiers (<code>-excl-variants-matching SNPID~rs1%</code>)
						or by any of the identifiers (<code>-excl-variants-matching rs1%</code>).
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Compute per-sample summary statistics:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -sample-stats -osample example.sample-stats
					</div>
				</div>

				<div class="task">
					<div class="task_name">
						Compute per-SNP summary statistics
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -snp-stats -osnp example.snp-stats
					</div>
					<div class="task_notes">
						Note: this computes snp summary statistics using all samples.  To use a subset of samples use the sample subsetting options, e.g.:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -snp-stats -osnp example.snp-stats -incl-samples-where sex=MALE
					</div>
					<div class="task_notes">
						where <code>sex</code> is a column in the sample file.
					</div>
				</div>

				<div class="task">
					<div class="task_name">
						Add genotype dosages for a given SNP or SNPs to the sample file
					</div>
					<div class="task_command_line">
						$ qctool -g example_#.gen -s example.sample -os example.sample -condition-on rs1234
					</div>
					<div class="task_notes">
						This adds a column with name <span class="monospace">rs1234:additive_dosage</span>, containing the additive dosage from the SNP, to the sample file.
						You can also select SNPs by position:
					</div>
					<div class="task_command_line">
						$ qctool -g example_#.gen -s example.sample -os example.sample -condition-on pos~03:10001
					</div>
					<div class="task_notes">
						It is also possible to select dominant, recessive, or heterozygote dosages from the SNP:
					</div>
					<div class="task_command_line">
						$ qctool -g example_#.gen 
						-s example.sample 
						-os example.sample 
						-condition-on "pos~03:10001(add|dom|het|rec)"
					</div>
					<div class="task_notes">
						Note: these options behave in the same way as SNPTEST's -condition_on option.
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Quantile-normalise columns of the sample file
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -os example.sample -quantile-normalise column1
					</div>
					<div class="task_notes">
						This adds a column with name <span class="monospace">column1:quantile-normalised</span> to the sample file.
						The specified columns must be continuous, i.e. of type 'P' or 'C'.
						Note that even though the SNPs are not used here, currently you must still specify the -g option.
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -os example.sample -quantile-normalise column1,column2
					</div>
					<div class="task_notes">
						Note: this option behaves in the same way as SNPTEST's -normalise option.
					</div>
				</div>
			</div>
		</section>

		<nav class="button_bar">
			<div class="nav_button" name="overview">
				<a href="../../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../../documentation/alphabetical_options.html">List of options</a></li>
					<li><a href="../../documentation/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/conversions.html">Converting between formats</a></li>
					<li><a href="../../documentation/mapping_and_aligning.html">Remapping and aligning data</a></li>
					<li><a href="../../documentation/annotations.html">Annotating data</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../../documentation/examples/combining.html">Combining datasets</a></li>
					<li><a href="../../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../../documentation/examples/filtering_samples.html">Filtering samples</a></li>
					<li><a href="../../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
<li><a href="../../documentation/examples/computing_pcs.html">Computing principal components</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../../documentation/download.html">download</a>
			</div>
		</nav>
		
	</body>
</html>
