<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../../js/local.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="../../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
			<h2>Assessing linkage disequilibrium</h2>
				<div class="task">
					<div class="task_name">
						Computing LD metrics
					</div>
					<div class="task_notes">
						The <code>-compute-ld-with</code> option can be used to compute LD metrics between
						the genotype data and a seperate set of genotype contained in a second file.  E.g.:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -compute-ld-with second.bgen second.sample -old sqlite://ld.sqlite:LD
					</div>
					<div class="task_notes">
						<p>
							For each biallelic variant in <code>example.bgen</code>, and each biallelic variant in <code>second.bgen</code>,
							QCTOOL constructs the pairwise table of non-missing genotypes, uses an EM algorithm to resolve phase
							of the double heterozygotes, and then outputs the frequency of each haplotype and the <em>D'</em>
							and <em>r<sup>2</sup></em> statistics.  (Multiallelic variants are currently not handled.)
					</p>
					<p>
						<b>Note:</b> QCTOOL outputs shrinkage estimates of LD by default.  This is done by augmenting the data
						with four 'prior' datapoints encoding the four possible pairwise haplotypes.  Roughly speaking,
						this corresponds to including the prior assumptions that 1: both variants are polymorphic
						and 2: there is at least some recombination between them.
						A practical effect of this is to shrink LD estimates toward zero.  This prior is weak
						and will only play a noticeable role when the sample size is very small.
					</p>
					<p>
						<b>Note 1</b>: to compute LD, samples are matched between datasets using the primary ID (the first column
						in the sample files) by default.  To alter this, use the <code>-match-sample-ids</code> option.
						This is described further on the <a href="merging.html">page on merging datasets</a>.
					</p>
					<p>
						<b>Note 2</b>:
						Currently LD output is always to a sqlite database file.  In the above command,
						results are placed in the file <code>ld.sqlite</code> and in a table called
						<code>LD</code>; a convenience view, called <code>LDView</code> is also constructed.  The simplest way
						to view that data is to use the <a href="https://sqlite.org/cli.html">sqlite3 command-line client<a>
							(which is already installed on most linux systems by default).  E.g. the command:
						</p>
					</div>
						<div class="task_command_line">
							$ sqlite3 -csv -header ld.sqlite "SELECT * FROM LDView LIMIT 10"
						</div>
						<div class="task_notes">
							<p>
								Will list the first ten LD records in CSV format.  The file can equally be accessed programmatically,
								for example using the <a href="https://docs.python.org/2/library/sqlite3.html">sqlite3 module</a> in python, or the <a href="https://cran.r-project.org/web/packages/RSQLite/index.html">RSQLite</a> package in R.
							</p>
							<p>
								We may add flat file support in future.
							</p>
						</div>
					</p>
					</div>
				</div>
				
				<div class="task">
				<div class="task_name">
					Restricting what is output
				</div>
				<div class="task_notes">
					Pairwise LD computations can generate a lot of data very fast.  The <code>-min-r2</code> and <code>-max-ld-distance</code>
					options can be used to restrict what is output.  E.g.:
				</div>
				<div class="task_command_line">
					$ qctool -g example.bgen -s example.sample -compute-ld-with second.bgen second.sample -old sqlite://ld.sqlite:LD -min-r2 0.05 -max-ld-distance 1Mb
				</div>
				<div class="task_notes">
				<p>
					This command only computes LD for variants on the same chromosome and within 1Mb of each other, and
					only outputs LD metrics when the computed r<sup>2</sup> value is at least 0.05.
				</p>
			</div>

				<div class="task">
					<div class="task_name">
						Stratifying LD computations across subsets
					</div>
					<div class="task_notes">
						As for <code>-snp-stats</code>, the <code>-stratify</code> option tells QCTOOL to compute LD statistics stratified over
						subsets of the data. E.g. suppose the sample file contains a column called <code>POP</code>
						which reflects the population group of each sample.  Then the command:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -compute-ld-with second.bgen second.sample -old sqlite://ld.sqlite:LD -stratify POP
					</div>
					<div class="task_notes">
						<p>
							will compute LD statistics for all pairs of variants in each population.  Output columns will be named
							in the form <code>POP=&lt;level&gt;:&lt;variable&gt;</code>.
						</p>
						<p>
							For example, suppose we've extracted data for the <a href="">O blood group mutation</a> rs8176719 from the
							<a href="">1000 Genomes Project</a> data into a file named <code>rs8176719.vcf</code>.  This command will
								find all tagging SNPs in the flanking region:
						</p>
					</div>
						<div class="task_command_line">
							$ qctool -g ALL.chr9.phase3_shapeit2_mvncall_integrated_v5.20130502.genotypes.vcf.gz
							-s 1000G.sample
							-compute-ld-with rs8176719.vcf 1000G.sample
							-old sqlite://ld.sqlite:LD
							-stratify POP
							-min-r2 0.75
							-max-distance 1Mb
						</div>
						<div class="task_notes">
						<p>
							Note that currently it's first necessary to make a sample file in the
							format <a href="../sample_file_formats.html">described here</a> for this to work.
						</p>
						<p>
							<b>Note:</b> This will also take a while to run, since at present it scans the whole of chromosome 9.
							A quicker way would be to use <a href="https://bitbucket.org/gavinband/bgen">bgenix</a> together with QCTOOL in 
							 <a href="pipelining.html">in a pipeline</a>.
						</p>
					</div>
				</div>

		</section>

		<nav class="button_bar">			<div class="nav_button" name="overview">
				<a href="../../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../../documentation/alphabetical_options.html">List of options</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../../documentation/examples/basic.html">Basic usage</a></li>
					<li><a href="../../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../../documentation/examples/combining.html">Combining datasets</a></li>
					<li><a href="../../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../../documentation/examples/filtering_samples.html">Filtering samples</a></li>
					<li><a href="../../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/examples/annotating.html">Computing annotations</a></li>
					<li><a href="../../documentation/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
					<li><a href="../../documentation/examples/computing_ld.html">Computing LD</a></li>
					<li><a href="../../documentation/examples/computing_pcs.html">Computing principal components</a></li>
					<li><a href="../../documentation/examples/inferring_ploidy.html">Working with sex chromosomes</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../../documentation/download.html">download</a>
			</div>
</nav>
		
	</body>
</html>
