<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../../js/local.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="../../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
			<h2>Computing principal components</h2>
			<p>
				Suppose <em>X</em> is a matrix of genotypes at a set of <em>L</em> biallelic
				variants (rows) for a collection of <em>N</em> samples (columns).
				We are interested in finding the eigenvectors of <em>1/N X<sup>t</sup>X</em>
				(the principal components)
				and of <em>1/L X X<sup>t</sup></em> (the 'loadings').
				The two are related by the singular value decomposition, which says that
				<em>X = U D V<sup>t</sup></em>
				where the columns of V<sup>t</sup> and the rows of U are the required eigenvectors, and <em>D</em>
				is a diagonal matrix.
			</p>
			<p>
				In practice principal components analysis (PCA) is usually carried out on the
				matrix of genotypes scaled by frequency, i.e.
				<em>x<sub>ij</sub> = g<sub>ij</sub> / &radic; f(1-f)</em>.  One reason for this is that
				(in simple theoretical situations) the rate of genetic drift is proportional
				to <em>&radic;f(1-f)</em>, where <em>f</em> is the ancestral allele frequency.
				If PCA is carried out on a sample from a set of diverged populations, the overall
				allele frequency can be though of as approximating the ancestral frequency, and the
				principal components will measure genetic drift between samples.
			</p>
			<p>
				The matrix <em>Z = 1/N X<sup>t</sup>X</em> can be thought of as a 'relatedness' matrix and is
				useful in its own right.  In simple theoretical scenarious, with the scaling <em>&radic;f(1-f)</em>,
				the matrix <em>Z</em> is an estimate of the 
				matrix of <a href="https://en.wikipedia.org/wiki/Coefficient_of_relationship">kinship
				coefficients</a>.  In practical situations <em>Z</em> is best thought of as providing a measure
				of relatedness relative to the sample; it can nevertheless be useful in identifying
				close relationships.
			</p>
			<p>
				QCTOOL computes principal components by solving for eigenvectors of <em>1/N X<sup>t</sup>X</em>.
				It computes loadings by projecting genotypes at each variant onto
				those eigenvectors, namely as <em>V<sup>t</sup></em> = U<sup>-1</sup> D<sup>-1</sup> X</em>.
			</p>

				<div class="task">
					<div class="task_name">
						Computing a kinship matrix
					</div>
					<div class="task_notes">
						The <code>-kinship</code> option can be used to compute a kinship matrix, e.g.:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -kinship kinship.csv
					</div>
					<div class="task_notes">
						This outputs pairwise kinship values to the file <code>kinship.csv</code>, which is stored in a 'long' format
						with columns holding the first sample id, second sample id, the number of pairwise non-missing genotypes,
						and the estimated kinship value.
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Eigendecomposing a kinship matrix and computing PCs
					</div>
					<div class="task_notes">
						The <code>-UDUT</code> option can be used to compute a UDUT decomposition of the computed kinship matrix.
							E.g.
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -kinship kinship.csv -UDUT udut.csv
					</div>
					<div class="task_notes">
						This computes a UDUT decomposition of the kinship matrix.  To additionally output principal components (PCs),
						additionally add the <code>-PCs</code> option:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -kinship kinship.csv -UDUT udut.csv -PCs 20 -osample PCs.csv
					</div>
					<div class="task_notes">
					<p>
						The argument is the number of PCs to output.
					</p>
					<p>
						<b>Note:</b> PCs are output to the file specified by <code>-osample</code>.
						Specifying both <code>-sample-stats</code> and <code>-PCs</code> leads to a file that contains both
						per-sample summary statistics and PCs.
						See the <a href="../../documentation/summary_file_formats.html">page
							on summary statistic file formats</a> for more information on the format of the output.
					</p>
					<p>
						In some contexts it may be preferable to load a previously
						computed kinship matrix, rather than to recompute a new one. This can be acheived with the
						<code>-load-kinship</code> option:
					</p>
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -load-kinship kinship.csv -UDUT udut.csv -PCs 20 -osample PCs.csv
					</div>
				</div>

				<div class="task">
					<div class="task_name">
						Computing PC loadings
					</div>
					<div class="task_notes">
						Having computed the UDUT decomposition, per-variant loadings can be
						computed using the <code>-loadings</code> option, e.g.:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -load-udut udut.csv -loadings loadings.csv
					</div>
					<div class="task_notes">
						The <code>-nPCs</code> option can again be used to adjust how many loadings are computed.
						(You should ensure the same set of variants is used to compute
						loadings as were used in constructing the kinship matrix).
					</div>
				</div>

				<div class="task">
					<div class="task_name">
						Projecting additional samples onto a principal components decomposition
					</div>
					<div class="task_notes">
						Given a set of loadings, additional samples can be projected onto the loadings using
						the <code>-project-onto</code> option:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -project-onto loadings.csv projection.sample
					</div>
					<div class="task_notes">
						The first argument is a set of loadings to load, while the second argument is the output file to compute.
					</div>
				</div>
		</section>

		<nav class="button_bar">
			<div class="nav_button" name="overview">
				<a href="../../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../../documentation/alphabetical_options.html">List of options</a></li>
					<li><a href="../../documentation/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/conversions.html">Converting between formats</a></li>
					<li><a href="../../documentation/mapping_and_aligning.html">Remapping and aligning data</a></li>
					<li><a href="../../documentation/annotations.html">Annotating data</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../../documentation/examples/basic.html">Basic usage</a></li>
					<li><a href="../../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../../documentation/examples/combining.html">Combining datasets</a></li>
					<li><a href="../../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../../documentation/examples/filtering_samples.html">Filtering samples</a></li>
					<li><a href="../../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
					<li><a href="../../documentation/examples/computing_pcs.html">Computing principal components</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../../documentation/download.html">download</a>
			</div>
		</nav>
		
	</body>
</html>
