<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<script type="text/javascript" src="../../js/jquery-1.4.2.min.js"></script>
		<script type="text/javascript" src="../../js/local.js"></script>
		<!-- Google Analytics tracking code-->
		<script type="text/javascript">
			var _gaq = _gaq || [];
			_gaq.push(['_setAccount', 'UA-16521993-1']);
			_gaq.push(['_trackPageview']);

			(function() {
				var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
				ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
				var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
			})();
		</script>
		<link href="../../style/css/default.css" rel="stylesheet" type="text/css">
		<title></title>
	</head>
	<body>
		<div class="header" id="header1">
			<!--<img src = "style/images/qctool.png" class = "header_img" ></img>-->
			<div class="header_text">
				qctool v2
			</div>
			<div class="header_subtext">
				A tool for quality control and analysis of gwas datasets.
			</div>
		</div>

		<section class="main_text">
			<h2>Computing principal components</h2>
				<div class="task">
					<div class="task_name">
						Computing a kinship matrix
					</div>
					<div class="task_notes">
						The <code>-kinship<code> option can be used to compute a kinship matrix, e.g.:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -kinship kinship.csv
					</div>
					<div class="task_notes">
						This outputs pairwise kinship values to the file <code>kinship.csv</code>, which is stored in a 'long' format
						of sample 1 id, sample 2 id, the number of pairwise non-missing genotypes, and the kinship value.
					</div>
				</div>
				<div class="task">
					<div class="task_name">
						Eigendecomposing a kinship matrix and computing PCs
					</div>
					<div class="task_notes">
						The <code>-UDUT</code> option can be used to compute a UDUT decomposition of the computed kinship matrix.
							E.g.
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -kinship kinship.csv -UDUT udut.csv
					</div>
					<div class="task_notes">
						This computes a UDUT decomposition of the kinship matrix.  To output a sample file containing PCs,
						additionally add the <code>-PCs</code> option:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -kinship kinship.csv -UDUT udut.csv -PCs pcs.sample
					</div>
					<div class="task_notes">
					<p>
						By default 10 PCs are computed.  Use <code>-nPCs &lt;n&gt;</code> to specify a different number.
					</p>
					<p>
						In some contexts (e.g. when recomputing PCs on a subset of samples) it may be preferable to load a previously
						computed kinship matrix (or a previously computed UDUT decomposition), rather than to recompute a new one.
						This can be acheived with the
						<code>-load-kinship</code> and <code>-load-UDUT</code> options:
					</p>
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -load-kinship kinship.csv -UDUT udut.csv
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -load-UDUT udut.csv -nPCs 10 -PCs pcs.sample
					</div>
				</div>

				<div class="task">
					<div class="task_name">
						Computing PC loadings
					</div>
					<div class="task_notes">
						Having computed the UDUT decomposition, per-variant loadings can be
						computed using the <code>-loadings</code> option, e.g.:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -load-udut udut.csv -loadings loadings.csv
					</div>
					<div class="task_notes">
						The <code>-nPCs</code> option can again be used to adjust how many loadings are computed.
						(You should ensure the same set of variants is used to compute
						loadings as were used in constructing the kinship matrix).
					</div>
				</div>

				<div class="task">
					<div class="task_name">
						Projecting additional samples onto a principal components decomposition
					</div>
					<div class="task_notes">
						Given a set of loadings, additional samples can be projected onto the loadings using
						the <code>-project-onto</code> option:
					</div>
					<div class="task_command_line">
						$ qctool -g example.bgen -s example.sample -project-onto loadings.csv projection.sample
					</div>
					<div class="task_notes">
						The first argument is a set of loadings to load, while the second argument is the output file to compute.
					</div>
				</div>
		</section>

		<nav class="button_bar">
			<div class="nav_button" name="overview">
				<a href="../../index.html">overview</a>
			</div>
			<div class="nav_button" name="documentation">
				documentation
				<ul class="menu">
					<li><a href="../../documentation/diagram_of_options.html">Diagram of options</a></li>
					<li><a href="../../documentation/alphabetical_options.html">List of options</a></li>
					<li><a href="../../documentation/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/conversions.html">Converting between formats</a></li>
					<li><a href="../../documentation/mapping_and_aligning.html">Remapping and aligning data</a></li>
					<li><a href="../../documentation/annotations.html">Annotating data</a></li>
				</ul>
			</div>
			<div class="nav_button" name="tutorial">
				examples
				<ul class="menu">
					<li><a href="../../documentation/examples/basic.html">Basic usage</a></li>
					<li><a href="../../documentation/examples/converting.html">Converting file formats</a></li>
					<li><a href="../../documentation/examples/sorting.html">Sorting and reordering data</a></li>
					<li><a href="../../documentation/examples/combining.html">Combining datasets</a></li>
					<li><a href="../../documentation/examples/merging.html">Merging in variants</a></li>
					<li><a href="../../documentation/examples/altering_id_data.html">Altering variant identifying data</a></li>
					<li><a href="../../documentation/examples/pipelining.html">Using qctool in a pipeline</a></li>
					<li><a href="../../documentation/examples/filtering_samples.html">Filtering samples</a></li>
					<li><a href="../../documentation/examples/filtering_variants.html">Filtering variants</a></li>
					<li><a href="../../documentation/examples/summary_statistics.html">Computing summary statistics</a></li>
					<li><a href="../../documentation/examples/adding_to_sample_file.html">Adding data to a sample file</a></li>
					<li><a href="../../documentation/examples/computing_pcs.html">Computing principal components</a></li>
				</ul>
			</div>
			<div class="nav_button" name="file_formats">
				file formats
				<ul class="menu">
					<a href="../../documentation/genotype_file_formats.html">Genotype file formats</a>
					<a href="../../documentation/sample_file_formats.html">Sample file formats</a>
					<a href="../../documentation/summary_file_formats.html">Summary statistic file formats</a>
			</div>
			<div class="nav_button" name="download">
				<a href="../../documentation/download.html">download</a>
			</div>
		</nav>
		
	</body>
</html>
